cmd/crank/beta/render/runtime_docker.go

```diff
@@ -145,39 +154,80 @@ func GetRuntimeDocker(fn pkgv1beta1.Function, log logging.Logger) (*RuntimeDocke
 	}
 	r := &RuntimeDocker{
 		Image:      fn.Spec.Package,
+		Name:       "",
 		Cleanup:    cleanup,
 		PullPolicy: pullPolicy,
 		log:        log,
 	}
 	if i := fn.GetAnnotations()[AnnotationKeyRuntimeDockerImage]; i != "" {
 		r.Image = i
 	}
+	if i := fn.GetAnnotations()[AnnotationKeyRuntimeNamedContainer]; i != "" {
+		r.Name = i
+	}
 	return r, nil
 }
 
 var _ Runtime = &RuntimeDocker{}
 
-// Start a Function as a Docker container.
-func (r *RuntimeDocker) Start(ctx context.Context) (RuntimeContext, error) {
-	r.log.Debug("Starting Docker container runtime", "image", r.Image)
-	c, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
-	if err != nil {
-		return RuntimeContext{}, errors.Wrap(err, "cannot create Docker client using environment variables")
+func (r *RuntimeDocker) findContainer(ctx context.Context, cli *client.Client, containerName string, startContainer bool) (string, string) {
+	r.log.Debug("searching for Docker container", "name", r.Name)
+
+	filterArgs := filters.NewArgs()
+	filterArgs.Add("name", containerName)
+
+	containers, err := cli.ContainerList(ctx, container.ListOptions{
+		Filters: filterArgs,
+		All:     true, // Include stopped containers
+	})
+
+	if err == nil && len(containers) > 0 {
+		for _, Container := range containers {
+			// Check if the container is running
+			if Container.State == "running" {
+				r.log.Debug("reusing Docker container", "name", Container.Names, "ID", Container.ID, "image", Container.Image)
+				addr := fmt.Sprintf("%s:%d", Container.Ports[0].IP, Container.Ports[0].PublicPort)
+				return Container.ID, addr
+			}
+		}
+		if startContainer && containers[0].ID != "" {
```

Comment:
Similarly here - I believe returning early would keep this method less nested.

Reply:
dont see it here, its a loop ?